File: LeetCode/Python/0015. 3Sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        if n < 3:
            return []
        
        triplets = set()
        nums.sort()
        
        for i in range(n):
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            left, right = i + 1, n - 1
            
            while left < right:
                _sum = nums[left] + nums[right] + nums[i]
                
                if _sum == 0:
                    triplets.add((nums[i], nums[left], nums[right]))
                    left, right = left + 1, right - 1
                elif _sum < 0:
                    left += 1
                else:
                    right -= 1
                    
        return triplets

File: LeetCode/Java/0015. 3Sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        HashSet<List<Integer>> triplets = new HashSet<>();
        Arrays.sort(nums);
        int n = nums.length;
        int j, k, sum;
        
        for (int i = 0; i < n; i++) {
            if (i != 0 && nums[i] == nums[i - 1])
                continue;
            j = i + 1;
            k = n - 1;
            
            while (j < k) {
                sum = nums[i] + nums[j] + nums[k];
                if (sum == 0) {
                    List<Integer> triplet = Arrays.asList(nums[i], nums[j], nums[k]);
                    triplets.add(triplet);
                    j++;
                    k--;
                } 
                else if (sum > 0) 
                    k--;
                else if (sum < 0)
                    j++;
            }
        }
        return new ArrayList<List<Integer>>(triplets);
    }
}
